lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -317,21 +332,29 @@ def from_string(cls, collection_input, artifacts_manager, supplemental_signature
         return cls.from_requirement_dict(req, artifacts_manager)
 
     @classmethod
-    def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_options=True):
-        req_name = collection_req.get('name', None)
-        req_version = collection_req.get('version', '*')
-        req_type = collection_req.get('type')
+    def from_requirement_dict(
```

Comment:
This classmethod doesn't validate user input and I wish we would prioritize that. Refactoring it to add type hints this will cause merge conflicts for the pull request adding a much needed argument spec: https://github.com/ansible/ansible/pull/83715. I'd also prefer to see the code refactored in order to simplify it, instead of refactored only to add type hints. I think it is convoluted and difficult to read with or without the type hints.

Reply:
Yeah, I know. I have a few things I'd like to improve too. Although, I feel like having a bit more typing (even if incomplete), would help with other refactors too.

Reply:
I don't think we're gaining much from all the type casts here. I think it would be better to omit these changes to reduce merge conflicts with efforts being made to refactor this method.

Reply:
@mattclay if we leave this method untyped, it'll cascade into multiple violations throughout the patch. So I'd rather keep them.

Reply:
@mattclay I got rid of the casts and used an explicit `typing.Any` for the value but kept `str` for the key. I think this is ready for another review.